triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.oauth.middleware
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.ServerBinding
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.HttpCookie
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
+import com.daml.oauth.server.{Request => OAuthRequest, Response => OAuthResponse}
+import com.typesafe.scalalogging.StrictLogging
+import java.util.Base64
+import java.util.UUID
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.Try
+import spray.json._
+
+// This is an implementation of the trigger service authentication middleware
+// for OAuth2 as specified in `/triggers/service/authentication.md`
+object Server extends StrictLogging {
+  import com.daml.oauth.server.JsonProtocol._
+
+  private def toRedirectUri(uri: Uri) =
```

Comment:
I think we might want to make the authority here or probably even better just the full URL configurable. Consider a case where this is running behind NGINX. We might be listening on `localhost` but the external facing host will be different and the path could also be different (e.g. it could be behind some prefix).

Reply:
Good point